File: Problems/0460-lfu-cache.py
class Node(object):
    __slots__ = ("key", "val", "freq", "prev", "next")

    def __init__(self, key=0, val=0):
        self.key = key
        self.val = val
        self.freq = 1
        self.prev = None
        self.next = None


class DLinkedList(object):
    __slots__ = ("head", "tail", "size")

    def __init__(self):
        self.head = Node()
        self.tail = Node()
        self.head.next = self.tail
        self.tail.prev = self.head
        self.size = 0

    def add(self, node):
        node.next = self.head.next
        node.prev = self.head
        self.head.next.prev = node
        self.head.next = node
        self.size += 1

    def remove(self, node):
        node.prev.next = node.next
        node.next.prev = node.prev
        self.size -= 1

    def pop(self):
        if self.size == 0:
            return None
        node = self.tail.prev
        self.remove(node)
        return node


class LFUCache(object):

    def __init__(self, capacity):
        self.cap = capacity
        self.minFreq = 0
        self.nodes = {}
        self.freqs = {}

    def _update(self, node):
        f = node.freq
        lst = self.freqs[f]
        lst.remove(node)

        if f == self.minFreq and lst.size == 0:
            self.minFreq += 1

        node.freq += 1

        if node.freq not in self.freqs:
            self.freqs[node.freq] = DLinkedList()

        self.freqs[node.freq].add(node)

    def get(self, key):
        node = self.nodes.get(key)

        if node is None:
            return -1

        self._update(node)
        return node.val

    def put(self, key, value):
        if self.cap == 0:
            return

        node = self.nodes.get(key)

        if node:
            node.val = value
            self._update(node)
            return

        if len(self.nodes) == self.cap:
            node = self.freqs[self.minFreq].pop()
            del self.nodes[node.key]

        node = Node(key, value)
        self.nodes[key] = node

        if 1 not in self.freqs:
            self.freqs[1] = DLinkedList()

        self.freqs[1].add(node)
        self.minFreq = 1

File: Problems/0460-lfu-cache.cpp
#include <bits/stdc++.h>
using namespace std;

// LeetCode solution starts from here
class LFUCache {
    struct Node {
        int key, val, freq;
        Node *prev, *next;

        Node(int k = 0, int v = 0)
            : key(k), val(v), freq(1), prev(nullptr), next(nullptr) {}
    };

    struct DLinkedList {
        Node *head, *tail;
        int size;

        DLinkedList() {
            head = new Node();
            tail = new Node();
            head->next = tail;
            tail->prev = head;
            size = 0;
        }

        void add(Node* node) {
            node->next = head->next;
            node->prev = head;
            head->next->prev = node;
            head->next = node;
            size++;
        }

        void remove(Node* node) {
            node->prev->next = node->next;
            node->next->prev = node->prev;
            size--;
        }

        Node* pop() {
            if (!size)
                return nullptr;

            Node* node = tail->prev;
            remove(node);
            return node;
        }
    };

    int cap, minFreq;
    unordered_map<int, Node*> nodes;
    unordered_map<int, DLinkedList*> freqs;

    void update(Node* node) {
        int f = node->freq;
        DLinkedList* list = freqs[f];

        list->remove(node);

        if (f == minFreq && list->size == 0)
            minFreq++;

        node->freq++;

        if (!freqs.count(node->freq))
            freqs[node->freq] = new DLinkedList();

        freqs[node->freq]->add(node);
    }

public:
    LFUCache(int capacity) : cap(capacity), minFreq(0) {}

    int get(int key) {
        if (!nodes.count(key))
            return -1;

        Node* node = nodes[key];
        update(node);
        return node->val;
    }

    void put(int key, int value) {
        if (cap == 0)
            return;

        if (nodes.count(key)) {
            Node* node = nodes[key];
            node->val = value;
            update(node);
            return;
        }

        if (nodes.size() == cap) {
            Node* node = freqs[minFreq]->pop();
            nodes.erase(node->key);
            delete node;
        }

        Node* node = new Node(key, value);
        nodes[key] = node;

        if (!freqs.count(1))
            freqs[1] = new DLinkedList();

        freqs[1]->add(node);
        minFreq = 1;
    }
};